# Error Handling :
==================
As developer writting a code without an errors are very important, So today we will make some mistakes and will find, what errors we will getting.
 1) This below code is the correct code without any issues, It'll do addtion of num1 and num2.
     num1=eval(input("enter a number one: "))
     num2=eval(input("enter a number two: "))
     add=num1+num2  
     print(add) : output = num1(2)+num2(2) =4.

    Let's start the modifications here :
    ------------------------------------
    Error-1 : [Name-Error]
    ======================
      Instead of doing addition with num1+num2 I mistakenly do addtion with num3(which we not defined on the above code)        
        add=num1+num3 [Error: NameError: name 'number3' is not defined]
      Why we are getting this error :
      -------------------------------
        Because we didn't define the num3 variable. If you want fix this, please verify is that variable(num3) is defined on your code or not ?

    Error-2 : [Syntax-Error]
    ========================
     I intentionnaly remove the end bracket and will see what error we will get.
       num1=eval(input("enter a number one: ") & Run the code
       Error-Code : SyntaxError: '(' was never closed.
       Fix : Verify the syntax of your code, And add the correct synstax and run : num1=eval(input("enter a number one: "))

    Error-3 : [ZeroDivision-Error]
    ==============================
         number1=eval(input("enter a number one: "))
         number2=eval(input("enter a number two: "))
         add=number1/number2
         print(add) & run the code
       Inputs:
         enter a number one:  100
         enter a number two:  0
       
       Error :
       -------
          "ZeroDivisionError: division by zero"
       
       Why This error we will get :
       ----------------------------
          This ZeroDivisionError, which occurs when you try to divide a number by zero. 
          The traceback indicates that the error happened on the line where we perform the division
          Python tries to compute 100/0, but since division by zero is not defined, Python raises a ZeroDivisionError.
   
       Example-scenario-for-this-error: 
       --------------------------------
         a) Basic Concept of Division:
         -----------------------------
             Division is like splitting something into equal parts. For example, if you have 10 apples and want to share them equally among 2 friends, each friend gets 5 apples.
         b) Division by Zero:
         --------------------
             Now, imagine you want to divide 10 apples among 0 friends. Since there are no friends to share the apples with, it doesn’t make sense.You can’t divide something if there’s no one to receive it.
             In mathematical terms, dividing by zero doesn’t produce a meaningful or defined result. It’s like trying to answer “How many times does 0 fit into 10?” There’s no answer because dividing by zero breaks the rules of arithmetic.

    Error-4 : [ZeroDivision-Error]
    ==============================
      Code :  "Trying to do division with string with intiger", Run this code and see what error we will get.
      ------
          number1=(input("enter a number one: "))      #string
          number2=float(input("enter a number two: ")) #int
          add=number1/number2                          # Now you wanna do division with string with int
          print(add) & RUN.
      Provide the inputs :
     ---------------------
        enter a number one:  5
        enter a number two:  2
      Error-Massege :
      ---------------
        TypeError: unsupported operand type(s) for /: 'str' and 'int'
      Why it happens : 
      ----------------
        Because it doesn't support dividing a string (number1) by an integer (number2).
      Fix :
      ------
        To fix this, you should convert number1 to an integer or float before performing the division:
           number1 = float(input("enter a number one: "))  # convert to float

    Error-5 : [SyntaxError: incomplete input]
    ==========================================
      Code :  "Removed the closing bracket for print statement", Run this code and see what error we will get.
      ------
          number1=(input("enter a number one: "))      
          number2=float(input("enter a number two: "))
          add=number1/number2                       
          print(add
      Error : 
      -------
          print(add
             ^
          SyntaxError: incomplete input [The error occurs because the "print statement is missing a closing parenthesis"]
      Fix : 
      ------
          print(add)  # Added the closing parenthesis

    Error-5 : [SyntaxError: Missing parentheses in call to 'print'. Did you mean print(...)?]
    =========================================================================================
      Code :  "Removed the brackets for add in print statement", Run this code and see what error we will get.
      ------
          number1=(input("enter a number one: "))      
          number2=float(input("enter a number two: "))
          add=number1/number2                       
          print add
       Error :
       -------
          SyntaxError: Missing parentheses in call to 'print'. Did you mean print(...)?
        Fix :
        -----
           The correct syntax for Python 3 is print(add) with parentheses


# Exception Handling :
======================
     Take a scenario :
     ------------------
       Suppose we have written 500 lines of code, But we are getting error at 200 line, But This error is might not that much impact our code, but we still want to continue executing the remaining code.
     How :
     -----
       We can use exception handling to manage this situation. The goal is to isolate the problematic code so that it doesn’t affect the execution of the rest of the program.





      


      
